File: app/src/main/java/cc/lixiaoyu/wanandroid/core/search/SearchActivity.java
package cc.lixiaoyu.wanandroid.core.search;

import android.content.DialogInterface;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.Toolbar;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.zhy.view.flowlayout.FlowLayout;
import com.zhy.view.flowlayout.TagAdapter;
import com.zhy.view.flowlayout.TagFlowLayout;

import java.util.List;

import butterknife.BindView;
import cc.lixiaoyu.wanandroid.R;
import cc.lixiaoyu.wanandroid.base.MVPBaseSwipeBackActivity;
import cc.lixiaoyu.wanandroid.entity.HotKey;
import cc.lixiaoyu.wanandroid.core.detail.ArticleDetailActivity;
import cc.lixiaoyu.wanandroid.core.search.result.SearchResultActivity;

public class SearchActivity extends MVPBaseSwipeBackActivity<SearchContract.Presenter>
        implements SearchContract.View{

    @BindView(R.id.search_toolbar)
    Toolbar mToolbar;
    @BindView(R.id.search_et_wordinput)
    EditText mEtInput;
    @BindView(R.id.search_btn_search)
    Button mBtnSearch;
    @BindView(R.id.search_btn_clearHistory)
    Button mBtnClearHistory;

    @BindView(R.id.search_tf_history)
    TagFlowLayout mTfHistory;
    @BindView(R.id.search_tf_hotkeys)
    TagFlowLayout mTfHotkeys;
    @BindView(R.id.search_tf_commonsite)
    TagFlowLayout mTfCommonSite;

    @Override
    protected void initData() {
        mPresenter.start();
    }

    @Override
    protected void initView() {
        //设置返回按钮
        setSupportActionBar(mToolbar);
        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeButtonEnabled(true);
        //设置搜索按钮的点击事件
        mBtnSearch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String word = mEtInput.getText().toString().trim();
                if(!TextUtils.isEmpty(word)){
                    mPresenter.searchArticle(word);
                }
                mEtInput.setText("");
            }
        });
        //设置清空历史记录按钮的点击事件
        mBtnClearHistory.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showClearHistoryDialog();
            }
        });
    }

    /**
     * 显示清空历史记录的弹窗
     */
    private void showClearHistoryDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(SearchActivity.this);
        AlertDialog dialog = builder.setTitle("提示")
                .setMessage("确认清空历史记录吗")
                .setCancelable(true)
                .setPositiveButton("确定", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        mPresenter.clearHistory();
                    }
                })
                .setNegativeButton("取消", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                })
                .create();
        dialog.show();
    }

    @Override
    protected int attachLayout() {
        return R.layout.activity_search;
    }

    @Override
    protected SearchPresenter createPresenter() {
        return new SearchPresenter();
    }

    @Override
    public void showLoading() {

    }

    @Override
    public void hideLoading() {

    }

    @Override
    public void showSearchHistory(final List<String> histroyWords) {
        mTfHistory.setAdapter(new TagAdapter<String>(histroyWords){

            @Override
            public View getView(FlowLayout parent, int position, String historyWord) {
                TextView tv = (TextView) LayoutInflater
                        .from(SearchActivity.this)
                        .inflate(R.layout.layout_nav_item, mTfHistory, false);
                tv.setText(historyWord);
                return tv;
            }
        });
        mTfHistory.setOnTagClickListener(new TagFlowLayout.OnTagClickListener() {
            @Override
            public boolean onTagClick(View view, int position, FlowLayout parent) {
                String historyWord = histroyWords.get(position);
                showSearchResult(historyWord);
                return true;
            }
        });
    }

    @Override
    public void showHotkey(final List<HotKey> hotKeys) {
        mTfHotkeys.setAdapter(new TagAdapter<HotKey>(hotKeys){

            @Override
            public View getView(FlowLayout parent, int position, HotKey hotKey) {
                TextView tv = (TextView) LayoutInflater
                        .from(SearchActivity.this)
                        .inflate(R.layout.layout_nav_item, mTfHotkeys, false);
                tv.setText(hotKey.getName());
                return tv;
            }
        });
        mTfHotkeys.setOnTagClickListener(new TagFlowLayout.OnTagClickListener() {
            @Override
            public boolean onTagClick(View view, int position, FlowLayout parent) {
                String hotKey = hotKeys.get(position).getName();
                showSearchResult(hotKey);
                return true;
            }
        });
    }

    @Override
    public void showCommonSite(final List<WebSite> webSites) {
        mTfCommonSite.setAdapter(new TagAdapter<WebSite>(webSites){

            @Override
            public View getView(FlowLayout parent, int position, WebSite webSite) {
                TextView tv = (TextView) LayoutInflater
                        .from(SearchActivity.this)
                        .inflate(R.layout.layout_nav_item, mTfCommonSite, false);
                tv.setText(webSite.getName());
                return tv;
            }
        });
        mTfCommonSite.setOnTagClickListener(new TagFlowLayout.OnTagClickListener() {
            @Override
            public boolean onTagClick(View view, int position, FlowLayout parent) {
                WebSite webSite = webSites.get(position);
                ArticleDetailActivity.actionStart(SearchActivity.this,
                        webSite.toDetailParam());
                return true;
            }
        });
    }

    @Override
    public void showSearchResult(String keyword) {
        //打开搜索结果页
        SearchResultActivity.actionStart(this, keyword);
        //关闭当前页面
        finish();
    }
}

File: app/src/main/java/cc/lixiaoyu/wanandroid/core/search/SearchActivity.kt
package cc.lixiaoyu.wanandroid.core.search

import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import cc.lixiaoyu.wanandroid.adapter.FlowTagAdapter
import cc.lixiaoyu.wanandroid.core.detail.ArticleDetailActivity.Companion.actionStart
import cc.lixiaoyu.wanandroid.core.search.model.HotKey
import cc.lixiaoyu.wanandroid.core.search.model.WebSite
import cc.lixiaoyu.wanandroid.core.search.result.SearchResultActivity
import cc.lixiaoyu.wanandroid.databinding.ActivitySearchBinding
import com.google.android.flexbox.FlexboxLayoutManager
import kotlinx.coroutines.launch

class SearchActivity : AppCompatActivity() {

    private lateinit var viewModel: SearchViewModel
    private lateinit var binding: ActivitySearchBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProvider(this)[SearchViewModel::class.java]
        binding = ActivitySearchBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initView()
    }

    private fun initView() {
        //设置返回按钮
        setSupportActionBar(binding.toolbar)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setHomeButtonEnabled(true)
        }
        //设置搜索按钮的点击事件
        binding.btnSearch.setOnClickListener {
            val word = binding.wordInput.text.toString().trim { it <= ' ' }
            if (word.isNotEmpty()) {
                // 搜索
                viewModel.searchKeyword(word)
                showSearchResult(word)
            }
            binding.wordInput.setText("")
        }
        //设置清空历史记录按钮的点击事件
        binding.btnClearHistory.setOnClickListener {
            showClearHistoryDialog()
        }
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.hotKeyList.collect {
                    val hotKeyAdapter = FlowTagAdapter<HotKey>().apply {
                        setTitleConverter { data -> data.name }
                        setOnTagClickListener { view, position, data ->
                            val hotKeyName = data.name
                            viewModel.searchKeyword(hotKeyName)
                            showSearchResult(hotKeyName)
                        }
                    }
                    binding.tfHotkeys.apply {
                        adapter = hotKeyAdapter
                        layoutManager = FlexboxLayoutManager(this.context)
                    }
                    hotKeyAdapter.setData(it)
                }
            }
        }
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.commonSiteList.collect {
                    val siteNameAdapter = FlowTagAdapter<WebSite>().apply {
                        setTitleConverter { data -> data.name }
                        setOnTagClickListener { view, position, data ->
                            actionStart(this@SearchActivity, data.toDetailParam())
                        }
                    }
                    binding.tfCommonsite.apply {
                        adapter = siteNameAdapter
                        layoutManager = FlexboxLayoutManager(this.context)
                    }
                    siteNameAdapter.setData(it)
                }
            }
        }
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.searchHistoryList.collect {
                    val searchHistoryAdapter = FlowTagAdapter<String>().apply {
                        setTitleConverter { data -> data }
                        setOnTagClickListener { view, position, data ->
                            showSearchResult(data)
                        }
                    }
                    binding.tfHistory.apply {
                        adapter = searchHistoryAdapter
                        layoutManager = FlexboxLayoutManager(this.context)
                    }

                    searchHistoryAdapter.setData(it)
                }
            }
        }
    }

    /**
     * 显示清空历史记录的弹窗
     */
    private fun showClearHistoryDialog() {
        val builder = AlertDialog.Builder(this@SearchActivity)
        val dialog = builder.setTitle("提示")
            .setMessage("确认清空历史记录吗")
            .setCancelable(true)
            .setPositiveButton("确定") { _, _ -> viewModel.clearSearchHistory() }
            .setNegativeButton("取消") { dialog, _ -> dialog.dismiss() }
            .create()
        dialog.show()
    }

    private fun showSearchResult(keyword: String) {
        //打开搜索结果页
        SearchResultActivity.actionStart(this, keyword)
        //关闭当前页面
        finish()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

}